text/0111-auto-resource-detection.md

```diff
@@ -0,0 +1,110 @@
+
+Note there are some existing implementations of this already in the SDKs (see [below](#prior-art-and-alternatives)), but nothing currently in the specification.
+
+## Explanation
+
+In order to apply auto-detected resource information to traces & metrics, a user will need to:
+
+- Configure which resource detector(s) they would like to run (e.g. AWS EC2 detector)
+- Provide the resource information returned by the configured detector(s) to the relevant tracer or meter provider(s)
+
+Note this means that if the user wants to make use of auto-detected resource information, they will be required to explicitly pass the resource to the trace and/or metric provider(s).
+
+If multiple detectors are configured, and more than one of these successfully detects a resource, the resources will be merged according to the Merge interface already defined in the specification, i.e. the earliest matched resource's attributes will take precedence. Each detector may be run in parallel, but to ensure deterministic results, the resources must be merged in the order the detectors were added.
+
+A default implementation of a detector that reads resource data from the `OTEL_RESOURCE` environment variable will be included in the SDK. The environment variable will contain of a list of key value pairs, and these are expected to be represented in a format similar to the [W3C Correlation-Context](https://github.com/w3c/correlation-context/blob/master/correlation_context/HTTP_HEADER_FORMAT.md#header-value), i.e.: `key1=value1,key2=value2`. This detector must always be configured as the first detector by default.
+
+Custom resource detectors related to specific environments (e.g. specific cloud vendors) must be implemented outside of the SDK, and users will need to import these separately.
+
+In order to allow users to be able to use different detectors for different trace or metric providers, the detector(s) must be configured against a resource provider, and users can create multiple providers if desired.
+
+## Internal details
+
+As described above, the following will be added to the Resource SDK specification:
+
+- An interface for "detectors", to retrieve resource information
+- Specification for the resource provider, which can have custom detectors added, and will return a merged resource
+- Details of the default "From Environment Variable" detector implementation described above
+
+This is a relatively small proposal so is easiest to explain the details with a code example:
+
+### Usage
+
+The following example in Go creates a trace provider that uses resource information automatically detected from AWS or GCP:
+
+Assumes a dependency has been added on the `otel api`, `otel sdk`, `otel awsdetector`, and `otel gcpdetector` packages.
+
+```go
+rp := resource.NewMustProvider()
+rp.AddDetectors(awsdetector.EC2, gcpdetector.GCE, gcpdetector.GKE)
+resource := rp.AutoDetect(ctx)
+tp := sdktrace.NewProvider(sdktrace.WithResource(resource))
```

Comment:
I think it's good to be able to have it explicit like this but would be nice to have a simple out of the box default assuming autodetection would work based on above:

`sdktrace.NewProvider(sdktrace.DetectResource())` or `sdktrace.NewProvider(sdktrace.MustDetectResource())`

Reply:
I like this suggestion - it makes up for one of the trade-offs I'd mentioned below, and makes initialization much simpler.

I added your code below but allowing detectors to be provided as parameters (even if we do require configuration of detectors, this could still be initialized in one line). Depending on the outcome of the thread above, I may remove the extra parameters.